prov/efa/src/efa_fabric.c

```diff
@@ -479,11 +490,24 @@ static int efa_get_gdr_support(char *device_name)
 		goto out;
 	}
 
-	ret = fi_read_file(get_sysfs_path(), gdr_path, &gdr_support, sizeof(char));
-	if (ret < 0)
+	ret = fi_read_file(get_sysfs_path(), gdr_path,
+			   gdr_support_str, MAX_GDR_SUPPORT_STRLEN);
+	if (ret < 0) {
+		if (errno == ENOENT) {
+			/* sysfs file does not exist, gdr is not supported */
+			ret = 0;
+		}
+
 		goto out;
+	}
 
-	ret = atoi(&gdr_support);
+	if (ret == 0) {
+		EFA_WARN(FI_LOG_FABRIC, "Sysfs file %s is empty\n", gdr_path);
+		ret = -FI_EINVAL;
+		goto out;
+	}
+
+	ret = (gdr_support_str[0] == '1');
```

Comment:
This is not right.  It would return true if the value in the file was "10", or any string starting with 1.  Like I said in our chat, use `strncmp()`.  This should be `ret = (0 == strncmp(gdr_support_str, "1", MAX_GDR_SUPPORT_STRLEN));

Reply:
I see, thanks. Fixed in the new revision, also fixed the commit message.

Reply:
@bwbarrett  on second thought, maybe we should only compare number of bytes read from file? e.g.

``` strncmp(gdr_support_str, "1", ret)```

My concern is that bytes after ```ret``` in ```gdr_support_str``` (if any) has not been touched, and might not be '\0'.

Reply:
I don't think that will matter. strncmp will compare n bytes at most, so the 8 bytes you've statically allocated.

If we read 8 bytes from the file and it doesn't match "1" then the compare will fail even if the 8th byte is not '\0'.

Reply:
Thank you for your reply!

My concern is that initial content of ```gdr_support_str``` can be random, like  it can be ```"01234567"````

so when we read from file, the file content is correct, which is "1".

After read file, ```gdr_support_str``` became "11234567"

Now because we compare 8 bytes, the comparison result would be not equal, even though the file content is correct.

Reply:
Oh, I see your point now. Can we just memset `gdr_support_str` to zero to avoid that?

Reply:
memset is expensive. I think ```strncmp(gdr_support_str, "1", ret)``` should work. e.g. We only compare number of bytes we read from file. If file contains "10" we compare 2 bytes. If file contains "1", we compare 1 byte.

Reply:
We're worrying about a 8 byte memset in a one time initialization path?

Reply:
memset here sets off warning bells that we're depending on particular behaviors of `fi_read_file()`.  The right answer if we're worried about that case would be to make the length argument in `strncmp()` be `MIN(MAX_GDR_SUPPORT_STRLEN, ret)`.  Of course, I'd probably use a variable name like `read_len` rather than `ret`, but that's just because I like very verbose code and know that the compiler will optimize that all out anyway.

Reply:
Sure, make sense.